File: Unit_Testing_C++/src/ecommerce_platform.cpp
// ecommerce_platform.cpp
#include "ecommerce_platform.h"
#include "order.h"
#include <iostream>

ECommercePlatform::ECommercePlatform(std::string name)
    : name(name), order_counter(0) {}

void ECommercePlatform::add_product(Product product) {
    products.push_back(product);
}

void ECommercePlatform::add_customer(Customer customer) {
    customers.push_back(customer);
}

Order ECommercePlatform::create_order(int customer_id, std::vector<Product> products, std::vector<int> quantities) {
    Customer* customer = nullptr;
    std::vector<std::pair<Product, int>> productsMap;
    for (auto& cust : customers) {
        if (cust.get_customer_id() == customer_id) {
            customer = &cust;
            break;
        }
    }
    if (customer == nullptr || products.size()==0) {
        // Return a null order object if customer not found
      
        return Order(0, Customer(0, "", "", 0), {});
        //return NULL;
    }
    double total_cost=0.0;
    int order_counter = 0;
    for (size_t i = 0; i < products.size(); ++i) {
        int price = products[i].get_price();
        int quantity = quantities[i];
        if (products[i].get_stock() < quantity) {
            // Return a null order object if insufficient stock
            return Order(0, Customer(0, "", "", 0), {});
            //return NULL;
        }
        total_cost+= (price * quantity);
        products[i].update_stock(-quantity);
        productsMap.push_back({products[i],quantities[i]});
    }

    if (total_cost > customer->get_credit_limit()) {
        // Return a null order object if credit limit exceeded
        return Order(0, Customer(0, "", "", 0), {});
        //return NULL;
    }
    //orders.push_back(order);
    return Order(customer_id,* customer, productsMap);
    //return NULL;
}

File: Unit_Testing_Python/src/ecommerce_platform.py
# ecommerce_platform.py
from order import Order
class ECommercePlatform:
    def __init__(self, name):
        self.name = name
        self.products = []
        self.customers = []
        self.orders = []
        self.order_counter = 0

    def add_product(self, product):
        self.products.append(product)

    def add_customer(self, customer):
        self.customers.append(customer)

    def create_order(self, customer_id, product_ids, quantities):
        print("List of customers:", self.customers)  # Print list of customers
        print("Customer ID:", customer_id)  # Print customer ID
        customer = next((c for c in self.customers if c.customer_id == customer_id),None)
        if not customer:
            return "Customer not found."

        

        products = []
        for product_id, quantity in zip(product_ids, quantities):
            product = next((p for p in self.products if p.product_id == product_id), None)
            if product:
                if product.stock >= quantity:
                    products.append((product, quantity))
                    product.stock -= quantity
                else:
                    return f"Insufficient stock for product {product_id}. Available stock: {product.stock}"
            else:
                return f"Product with ID {product_id} not found."
        
        # Check if the total order cost exceeds the customer's credit limit
        total_cost = sum(self._get_product_price(product_id) * quantity for product_id, quantity in zip(product_ids, quantities))
        if total_cost > customer.credit_limit:
            return "Customer credit limit exceeded."

        if products:
            self.order_counter += 1
            order = Order(self.order_counter, customer, products)
            self.orders.append(order)
            return order
        else:
            return "No valid products in the order."

    def _get_product_price(self, product_id):
        product = next((p for p in self.products if p.product_id == product_id), None)
        return product.price if product else 0
